pkg/rancher-desktop/main/extensions/extensions.ts

```diff
@@ -0,0 +1,223 @@
+  id: string;
+  /** The directory this extension will be installed into */
+  readonly dir: string;
+  protected readonly client: ContainerEngineClient;
+  protected _metadata: Promise<ExtensionMetadata> | undefined;
+
+  /** Extension metadata */
+  get metadata(): Promise<ExtensionMetadata> {
+    this._metadata ??= (async() => {
+      const fallback = { vm: {} };
+
+      try {
+        const raw = await this.client.readFile(this.id, 'metadata.json');
+        const result = _.merge({}, fallback, JSON.parse(raw));
+
+        if (!result.icon) {
+          throw new ExtensionErrorImpl(ExtensionErrorCode.INVALID_METADATA, 'Invalid extension: missing icon');
+        }
+
+        return result;
+      } catch (ex: any) {
+        console.error(`Failed to read metadata for ${ this.id }: ${ ex }`);
+        // Unset metadata so we can try again later
+        this._metadata = undefined;
+        throw new ExtensionErrorImpl(ExtensionErrorCode.INVALID_METADATA, 'Could not read extension metadata', ex);
+      }
+    })();
+
+    return this._metadata as Promise<ExtensionMetadata>;
+  }
+
+  protected _iconName: Promise<string> | undefined;
+
+  /** iconName is the file name of the icon (e.g. icon.png, icon.svg) */
+  get iconName(): Promise<string> {
+    this._iconName ??= (async() => {
+      return `icon${ path.extname((await this.metadata).icon) }`;
+    })();
+
+    return this._iconName as Promise<string>;
+  }
+
+  async install(): Promise<boolean> {
+    const metadata = await this.metadata;
+
+    await fs.promises.mkdir(this.dir, { recursive: true });
+    try {
+      await this.installMetadata(this.dir, metadata);
+      await this.installIcon(this.dir, metadata);
+      await this.installUI(this.dir, metadata);
+      await this.installHostExecutables(this.dir, metadata);
+    } catch (ex) {
+      console.error(`Failed to install extension ${ this.id }, cleaning up:`, ex);
+      await fs.promises.rm(this.dir, { recursive: true }).catch((e) => {
+        console.error(`Failed to cleanup extension directory ${ this.dir }`, e);
+      });
+      throw ex;
+    }
+
+    mainEvents.emit('settings-write', { extensions: { [this.id]: true } });
+
+    // TODO: Do something so the extension is recognized by the UI.
+    console.debug(`Install ${ this.id }: install complete.`);
+
+    return true;
+  }
+
+  protected installMetadata(workDir: string, metadata: ExtensionMetadata): Promise<void> {
+    return fs.promises.writeFile(
+      path.join(workDir, 'metadata.json'),
+      JSON.stringify(metadata, undefined, 2));
+  }
+
+  protected async installIcon(workDir: string, metadata: ExtensionMetadata): Promise<void> {
+    try {
+      const origIconName = path.basename(metadata.icon);
+
+      try {
+        await this.client.copyFile(this.id, metadata.icon, workDir);
+      } catch (ex: any) {
+        throw new ExtensionErrorImpl(ExtensionErrorCode.FILE_NOT_FOUND, `Could not copy icon file ${ metadata.icon }`, ex);
+      }
```

Comment:
Oh, wait, this doesn't actually work: the exception is from `this.client.copyFile`, which happens in a VM; therefore, the exception never has a code of `ENOENT` (because it's `limactl shell` failing, it's just a `SpawnError` with code 1).